propulsion-reactor/Program.fs

```diff
@@ -351,29 +372,41 @@ module Args =
             | _, _, Some p, _ ->            Percentage p
             | None, None, None, true ->     StartPos.TailOrCheckpoint
             | None, None, None, _ ->        StartPos.StartOrCheckpoint
-
-        member __.Discovery =
-            match __.Tcp, __.Port with
-            | false, None ->   Discovery.GossipDns            __.Host
-            | false, Some p -> Discovery.GossipDnsCustomPort (__.Host, p)
-            | true, None ->    Discovery.Uri                 (UriBuilder("tcp", __.Host).Uri)
-            | true, Some p ->  Discovery.Uri                 (UriBuilder("tcp", __.Host, p).Uri)
         member __.Tcp =
-            a.Contains EsSourceParameters.Tcp
+            a.Contains Tcp
             || EnvVar.tryGet "EQUINOX_ES_TCP" |> Option.exists (fun s -> String.Equals(s, bool.TrueString, StringComparison.OrdinalIgnoreCase))
         member __.Port =                    match a.TryGetResult Port with Some x -> Some x | None -> EnvVar.tryGet "EQUINOX_ES_PORT" |> Option.map int
         member __.Host =                    a.TryGetResult Host     |> defaultWithEnvVar "EQUINOX_ES_HOST"     "Host"
         member __.User =                    a.TryGetResult Username |> defaultWithEnvVar "EQUINOX_ES_USERNAME" "Username"
         member __.Password =                a.TryGetResult Password |> defaultWithEnvVar "EQUINOX_ES_PASSWORD" "Password"
+        member __.ProjTcp =
+            a.Contains ProjTcp
+            || match EnvVar.tryGet "EQUINOX_ES_PROJ_TCP" with Some s -> String.Equals(s, bool.TrueString, StringComparison.OrdinalIgnoreCase) | None -> __.Tcp
+        member __.ProjPort =                match a.TryGetResult ProjPort with
+                                            | Some x -> Some x
+                                            | None -> (EnvVar.tryGet "EQUINOX_ES_PROJ_PORT" |> Option.map int) |> Option.orElse __.Port
```

Comment:
for consistency, orElse should be defaultWith (also orElse is prob fairly new?)

Reply:
getting value of __.Port doesn't fail with an exception and is an optional value. Yes, it is relatively new.

Reply:
There's `orElseWith` but we don't need it as __.Port won't throw an exception. I can change it for consistency to `orElseWith` but `defaultWith` won't help as it returns a non-optional value.

Reply:
I get that it's functionally the same and that it won't throw - I was suggesting
a) ~we should be conservative about orElse (it might not work on 2.1.500)~ orElseWith is worth taking a gamble on - we can always add a paste of https://github.com/fsharp/fsharp/blob/master/src/fsharp/FSharp.Core/option.fs#L29 in the helpers if it becomes a problem - I bet there are one or two other places in the templates where it can simplyify things. on reflection, I think assuming 3.1.101 should be considered the standard
b) to follow the same layout/scheme/impl pattern of the other fallbacks - i.e. use the `With` form even if not strictly necessary as this whole set of fns follow that pattern (doesnt mean I wouldnt still default to a match statement over a set of piped option functions if i think its easier to read, and its certainly my goto/default)
```
// as we have >1, the cleanup is now warranted (when there was one I was wearing the pain)
let tryEnvVarTrue varName = EnvVar.tryGet varName |> Option.exists (fun s -> String.Equals(s, bool.TrueString, StringComparison.OrdinalIgnoreCase)
...
member __.ProjTcp =
    a.Contains ProjTcp
    || (tryEnvVarTrue  "EQUINOX_ES_PROJ_TCP"
        |> Option.defaultWith (fun () -> __.Tcp))
member __.ProjPort =
    a.TryGetResult ProjPort
    || (EnvVar.tryGet "EQUINOX_ES_PROJ_PORT"
        |> Option.map int
        |> Option.OrElseWith (fun () -> __.Port))

Reply:
As discussed, will use below format and remove `defaultWithEnvVarOrFallback` function

```
match a.TryGetResult ProjHost with
| Some x -> x
| None -> EnvVar.tryGet "EQUINOX_ES_PROJ_HOST" |> Option.defaultWith (fun () -> __.Host)
```

where as for `tcp` it is not feasible to do `Option.defaultWith (fun () -> __.Tcp))`
in scenario where cluster es is `-T` but proj es isn't, it will always fallback proj es to be TCP. We can keep default value for tcp as false for both, and will only be considered as tcp flag is specified.

We need to think about the scenario where `-T` is specified but none of the proj arguments are provided, i.e. we need to fallback to cluster configurations by setting `tcp` flag for proj es.

Reply:
Good point re the tcp flag 👍